File: cpp/322.cpp
// 零钱兑换
// Coin Change

#include "common.h"
#include <algorithm>

struct Solution
{
    int coinChange(std::vector<int> &coins, int amount)
    {
        std::vector<int> minCoins(amount + 1, -1);
        minCoins[0] = 0;
        std::sort(coins.begin(), coins.end());

        for (int i = 0; i <= amount; i++)
        {
            for (int c : coins)
            {
                if (i + c <= amount && minCoins[i] != -1)
                {
                    if (minCoins[i + c] == -1 || minCoins[i + c] > minCoins[i] + 1)
                    {
                        minCoins[i + c] = minCoins[i] + 1;
                    }
                }
            }
        }
        return minCoins[amount];
    }
};

int main()
{
    Solution s;
    std::vector<std::tuple<std::vector<int>, int>> test_cases = {
        {{1, 2, 5}, 11},
        {{2}, 3},
        {{1}, 0},
    };
    for (auto &[coins, amount] : test_cases)
    {
        std::cout << s.coinChange(coins, amount) << std::endl;
    }
}

File: python/322.py
# 零钱兑换
# Coin Change

from common import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        minCoins = [-1] * (amount + 1)
        minCoins[0] = 0
        coins.sort()

        for i in range(amount + 1):
            for c in coins:
                if i + c <= amount and minCoins[i] != -1:
                    if minCoins[i + c] == -1 or minCoins[i + c] > minCoins[i] + 1:
                        minCoins[i + c] = minCoins[i] + 1

        return minCoins[amount]


if __name__ == "__main__":
    test_cases = [
        ([186, 419, 83, 408], 6249),
        ([1, 2, 5], 11),
        ([2], 3),
        ([1], 0),
    ]
    for coins, amount in test_cases:
        print(Solution().coinChange(coins, amount))
